.github/coding-style.md

```diff
@@ -0,0 +1,171 @@
+var (
+	cache = NewCache()
+)
+...
+func (w *widget) Update(x X) error {
+    w.mutex.Lock()
+	defer w.mutex.Unlock()
+	
+	if cache.Contains(x) {
+	    ...	
+    }
+	...
+}
+```
+
+```go
+// Better: `cache` users likely understand that manipulations require
+// synchronization might miss future changes to this contract. 
+var (
+	cache = NewCache()
+    mutex sync.RWMutex
+)
+...
+func (w *widget) Update(x X) error {
+    mutex.Lock()
+	defer mutex.Unlock()
+	
+	if cache.Contains(x) {
+	    ...	
+    }
+	...
+}
+```
+
+```go
+// Good: `cache` is only used by `widget` methods likely living in the same file.
+type widget struct {
+  cache Cache
+  mutex sync.RWMutex
+}
+...
+func (w *widget) Update(x X) error {
+    w.mutex.Lock()
+	defer w.mutex.Unlock()
+	
+	if w.cache.Contains(x) {
+	    ...	
+    }
+	...
+}
+```
+
+Capture in comments assumptions and decisions that are not obvious from the
+code, for example:
+```go
+var (
+	authorizer = perrpc.FromMap(map[authz.Authorizer][]string{
+		// When this endpoint was public, Sensor relied on it to check Central's
+		// availability. While Sensor might not do so today, we need to ensure
+		// backward compatibility with older Sensors.
+		or.SensorOr(user.Authenticated()): {
+			"/v1.MetadataService/GetMetadata",
+		},
+	})
+)
+```
+
+## Guidelines
+
+- Always use `defer mutex.Unlock()` instead of explicitly calling `Unlock()`.
+- A `Close()` on a file etc. should only be deferred and wrapped in
+  `utils.IgnoreError` if you exclusively _read_. If you write to a file, you
+  **must** check for a `nil` error upon close, otherwise you cannot be certain
+  that all data has been persisted.
+- Use the fact that certain operations return zero values by default to write
+  conditions in a more goal-oriented way. E.g., prefer
+  `if x := m[k]; x != nil { foo(*x) }` over `if x, ok := m[k]; ok { foo(*x) }`.
+  The former makes it apparent that the `*x` cannot panic while the latter only
+  after checking that `m` indeed can never contain `nil` values.
+
+### Declarations
+
+- Prefer `var x T` instead of `x := T{}` or `x := 0/false/…` for initializing
+  zero-valued local variables.
+- Avoid `var ( … )` blocks within function scopes, only use them at the global
+  scope.
+- Declarations at the global scope should follow this order: const block, var
+  block, everything else. If you define struct types and methods of this type in
+  the same file, the order should be: type declaration, new/constructor function,
+  methods. Do not interleave methods of different types.
+- If you are registering packages via `_` then extract this line into its own
+  import block and preface it with a comment explaining what this import does.
+- Scope variables as low as possible.
+- Only use `if` + assignment with `:=`.
+
+### Functions
+
+- Every function that can potentially block should receive a `ctx context.Context`
+  as the first parameter.
+- An exported function should only ever return exported types (even though Go
```

Comment:
David brought up an interesting point in a code review:

Assume we have an interface and it's implementation as such:
```go
type Reader interface {
   Read() string
}

type readerImpl struct {}

func (r *readerImpl) Read() string { return "hello world" }
```

Currently in most of our code in case we want to instantiate a read we do:
```go
func NewReader() Reader { return &readerImpl{} }
```

There are some Go best-practices which mention that returning a concrete type is preferred and clients can specify the interface.

This then would lead to things like:
```go
type Reader interface {
    Read() string
}

type ReaderImpl struct {}
```

where the implementation would be exposed as well. I feel this is suboptimal, so should we add something within the guidelines here to specify that we strive to return interfaces and hide concrete implementation types?